src/Main.vala

```diff
@@ -1,28 +1,56 @@
 using Gtk;
+using GtkLayerShell;
+
+// Globals
+bool IS_SESSION_WAYLAND;
+string WM_NAME;
 
 /**
  * Application entry point
  */
 public static int main (string[] args) {
     Gtk.init (ref args);
 
+    // Get session type (wayland or x11) and set the flag
+    string session_type = Environment.get_variable ("XDG_SESSION_TYPE");
+    string gdk_backend = Environment.get_variable ("GDK_BACKEND");
+    IS_SESSION_WAYLAND = session_type == "wayland" && gdk_backend != "x11";
+
+    // Set window manager
+    string sway_sock = Environment.get_variable ("SWAYSOCK");
+    string i3_sock = Environment.get_variable ("I3SOCK");
+
+    if(sway_sock != null) {
+        WM_NAME = "sway";
+    } else if(i3_sock != null) {
+        WM_NAME = "i3";
+    } else {
+        WM_NAME = "Unknown";
+    }
+
     var arg_map = parse_args (args);
     if (arg_map.contains ("-h") || arg_map.contains ("--help")) print_help_and_exit ();
 
     var window = new Ilia.DialogWindow (arg_map);
+    window.destroy.connect (Gtk.main_quit); 
+    
+    // Grab inputs from wayland backend before showing window
+    if(IS_SESSION_WAYLAND) {
+        GtkLayerShell.init_for_window(window);
+        GtkLayerShell.set_keyboard_mode(window, GtkLayerShell.KeyboardMode.EXCLUSIVE);
```

Comment:
question: Is this guaranteed to succeed?  In X11 it's not and it does often fail which is why there's some shenanigans in this code...

Reply:
to clarify, shenanigans in the code that **I wrote** that polls to grab keyboard/mouse.

Reply:
Yes, it is guaranteed to succeed for `wlroots` based window managers. `wlroots` provides a dedicated protocol for sticky windows called [`zwlr_layer_shell_v1`](https://wayland.app/protocols/wlr-layer-shell-unstable-v1) which allows applications to assign themselves the `surface_layer` role. This role tells the window manager to render the application in its own layer, within which the application can position it's windows as it desires. The protocol also provides a [method](https://wayland.app/protocols/wlr-layer-shell-unstable-v1#zwlr_layer_surface_v1:request:set_keyboard_interactivity) to define how keyboard events are delivered to the layer/application.

Reply:
`GTKLayerShell` is a GTK-based abstraction for interacting with the layer shell protocol. I'm using the `set_keyboard_mode` method is used to essentially assign the `ilia` window [`EXCLUSIVE`](https://wayland.app/protocols/wlr-layer-shell-unstable-v1#zwlr_layer_surface_v1:enum:keyboard_interactivity:entry:exclusive) keyboard interactivity, which means all inputs will be delivered only to `ilia`.